scl/compressors/lz77.py

```diff
@@ -128,6 +124,310 @@ class LZ77Sequence:
+
+            # Now encode the counts using Elias Delta code.
+            # We first generate a list of counts (of length self.alphabet_size) and then apply Elias Delta.
+            # For any bytes not seen in the data, set count to 0.
+            for i in range(self.alphabet_size):
+                if i not in counts:
+                    counts[i] = 0
+            counts_list = [counts[i] for i in range(self.alphabet_size)]
+
+            counts_encoding = EliasDeltaUintEncoder().encode_block(DataBlock(counts_list))
+            # combine everything into a single bitarray
+            return (
+                uint_to_bitarray(len(counts_encoding), ENCODED_BLOCK_SIZE_HEADER_BITS)
+                + counts_encoding
+                + uint_to_bitarray(len(values_encoding), ENCODED_BLOCK_SIZE_HEADER_BITS)
+                + values_encoding
+            )
+        else:
+            # if no counts (i.e., no values) just transmit 0
+            return uint_to_bitarray(0, ENCODED_BLOCK_SIZE_HEADER_BITS)
+
+
+class EmpiricalIntHuffmanDecoder(DataDecoder):
+    def __init__(self, alphabet_size):
+        self.alphabet_size = alphabet_size
+
+    def decode_block(self, encoded_bitarray: BitArray):
+        num_bits_consumed = 0
+        # first read the size of the counts (i.e., frequencies) encoding
+        counts_encoding_size = bitarray_to_uint(encoded_bitarray[:ENCODED_BLOCK_SIZE_HEADER_BITS])
+        num_bits_consumed += ENCODED_BLOCK_SIZE_HEADER_BITS
+        # now decode the counts using Elias Delta
+        if counts_encoding_size == 0:
+            return DataBlock([]), num_bits_consumed  # no  values
+        counts, num_bits_consumed_counts = EliasDeltaUintDecoder().decode_block(
+            encoded_bitarray[num_bits_consumed : num_bits_consumed + counts_encoding_size]
+        )
+        counts = counts.data_list
+        assert counts_encoding_size == num_bits_consumed_counts
+        num_bits_consumed += counts_encoding_size
+        # generate the probability distribution from the counts
+        counts = {i: counts[i] for i in range(self.alphabet_size) if counts[i] > 0}
+        prob_dist = ProbabilityDist.normalize_prob_dict(counts)
+
+        # load the encoding of the values
+        huffman_encoding_size = bitarray_to_uint(
+            encoded_bitarray[num_bits_consumed : num_bits_consumed + ENCODED_BLOCK_SIZE_HEADER_BITS]
+        )
+        num_bits_consumed += ENCODED_BLOCK_SIZE_HEADER_BITS
+        # decode values with Huffman
+        decoded_vals, num_bits_consumed_vals = HuffmanDecoder(prob_dist).decode_block(
+            encoded_bitarray[num_bits_consumed : num_bits_consumed + huffman_encoding_size]
+        )
+        assert huffman_encoding_size == num_bits_consumed_vals
+        num_bits_consumed += huffman_encoding_size
+        return decoded_vals, num_bits_consumed
+
+
+class LogScaleBinnedIntegerEncoder(DataEncoder):
+    """
+    Encodes a list of non-negative integers by binning in log scale and then
```

Comment:
The magnitude of difference (residual) can be between 0 - 2^logarithm, which can be a large range. Maybe comment on what kind of distributions is this encoding good. (exponential?)

Reply:
we actually use a fixed length code for that